File: android_project/folder_type/BadThreadExam/app/src/main/java/com/tenisme/badthreadexam/MainActivity.java
package com.tenisme.badthreadexam;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;

// Thread 사용의 안 좋은 예시를 보여주기 위한 프로젝트
public class MainActivity extends AppCompatActivity {

    private Thread thread;
    private int count = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    public void startThread(View view) {
        if(thread == null) {
            thread = new Thread("My Thread") {
                @Override
                public void run() {
                    for(int i = 0; i < 100; i++) {
                        try {
                            count++;
                            Thread.sleep(1000);
                        } catch (InterruptedException e) {
                            break;
                        }
                        Log.d("MyThread", "스레드 동작 중 : " + count);
                    }
                }
            };
            thread.start();
        }
    }
    
    public void stopThread(View view) {
        if(thread != null) {
            thread.interrupt();
            thread = null;
            count = 0;
        }
    }
}

File: android_project/folder_type/RoomTest/app/src/main/java/com/tenisme/roomtest/MainActivity.kt
package com.tenisme.roomtest

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ViewModelProvider
import com.tenisme.roomtest.databinding.ActivityMainBinding
import kotlinx.coroutines.*
import java.lang.Runnable

class MainActivity : AppCompatActivity() {

    private val binding by lazy {
        DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main)
    }

    private val memoViewModel by lazy {
        ViewModelProvider(this, ViewModelProvider.AndroidViewModelFactory(application))
            .get(MemoViewModel::class.java)
    }

    private var sentenceId: Long? = null
    private var sentences: List<MemoEntity>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        binding.lifecycleOwner = this
        binding.memoViewModel = memoViewModel

        memoViewModel.getAllMemoLive().observe(this, {
            Log.d("Test", "Observing Sentences : $it")
        })

        runBlocking {
            GlobalScope.launch(Dispatchers.IO) {
                sentenceId = memoViewModel.insert(MemoEntity("Test2"))
                Log.d("Test", "InsertedSentenceId : $sentenceId")

                val sentence = memoViewModel.getMemo(sentenceId!!)
                Log.d("Test", "Sentence in Scope : $sentence")
            }
        }
    }
}